### ISSUE TITLE
Recommended response for empty packets

### ISSUE THREAD

**Issue by huitema** [quicwg/base-drafts#1835]
Draft-15 says in section 5 that "_QUIC payloads MUST contain at least one frame, and MAY contain
multiple frames and multiple frame types._" That's fine, but there is no text specifying the receiver behavior. How about, "_Nodes who receive a packet with an empty payload MUST terminate the connection with error PROTOCOL_VIOLATION_"?

**Comment by mikkelfj** [comment]
That won't work in early handshake but it could make sense later on. Why have this restriction and complication in the first place rather than just allowing empty frames?

**Comment by nibanks** [comment]
I also disagreed with adding the extra complexity added on the receiver side. Personally, I'd prefer to add that the receiver MAY terminate the connection.

**Comment by mikkelfj** [comment]
Hmm, I'm not sure the protocol should be designed around babysitting special cases. You argument could just as well apply to packets with padding only or packets with only NEW_CONNECTION_ID when there are no support for changing CID's, etc.  It is already stated that a packet cannot be dropped without veryfying the AEAD tag.

**Comment by mikkelfj** [comment]
Empty packets also serve a keep-alive purpose. That said, there is another PR, or something, suggesting packets should have a minimum length to simplify PNE. That would naturally void empty packets.

**Comment by huitema** [comment]
My main point is that the spec says MUST NOT be empty without specifying receiver behavior. So the spec should either not say that, or specify the behavior.

**Comment by mikkelfj** [comment]
I agree with your point, I'm just mentioning that any resolution needs to consider handshake behavior in a way that does open for DoS injection.

**Comment by ianswett** [comment]
I sent out a PR with a variant of Christian's suggestion using MAY.

If one wants a keepalive, a packet containing a PING likely makes more sense than an empty one, since an empty one won't cause an ACK to be sent in response.

**Comment by igorlord** [comment]
If a client wants to keep its NAT binding alive, it may want to send a PADDING (or nothing at all, if it were supported).
